File: ELBackend/app/services/vocabulary_scheduler/scheduler.py
"""Vocabulary Scheduler — orchestrates pool building, scoring, and allocation for an ArcPlan."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from app.models.vocabulary import UserVocabulary
from app.services.vocabulary_scheduler.pools import apply_pending_overlay, build_pools
from app.services.vocabulary_scheduler.scorer import final_score, score_context
from app.services.vocabulary_scheduler.allocator import (
    allocate_main_episode,
    allocate_side_episode,
)


async def schedule(
    arc_plan: dict,
    user_vocab: dict,
    now: datetime | None = None,
    episode_limit: int = 10,
    llm_client: Any = None,
) -> dict:
    """Schedule vocabulary for all episodes in an arc plan.

    Fills each episode's target_words by:
    1. Building unseen and due-review pools from user vocabulary
    2. Applying pending word priority overlay
    3. For each episode: scoring candidates, computing final scores, allocating words

    Args:
        arc_plan: ArcPlan dict with "episodes" (list of episode dicts) and
                  "pending_words" (list of {item_id, rejected_count} dicts) keys.
                  Episode dicts have "source_text" and "episode_type" ("main"/"side").
        user_vocab: UserVocabulary dict with "vocabulary" key containing list of item dicts.
        now: Reference datetime. Defaults to UTC now if None.
        episode_limit: Max new words AND max review words per episode (default 10).
        llm_client: Optional LLM client for contextual scoring. When None, falls back to 0.5 for all candidates.

    Returns:
        Updated arc_plan dict with each episode's "target_words" list populated.
        target_words entries are TargetWord Pydantic models serialized via model_dump().
    """
    if now is None:
        now = datetime.now(timezone.utc)

    # Convert input dict to UserVocabulary model
    vocab = UserVocabulary.model_validate(user_vocab)

    pending_words: list[dict] = arc_plan.get("pending_words", [])
    episodes: list[dict] = arc_plan.get("episodes", [])

    # Build pools once, then apply pending overlay (both return list[VocabularyItem])
    pools = build_pools(vocab, now)
    unseen_pool, review_pool = apply_pending_overlay(pools, pending_words)

    # Track pool consumption positions (each episode takes the next batch)
    unseen_pos: int = 0
    review_pos: int = 0

    # Track arc-wide is_new dedup
    arc_new_ids: set[str] = set()

    for episode in episodes:
        source_text: str | None = episode.get("source_text")
        episode_type: str = episode.get("episode_type", "main")

        # Take next batch from each pool (up to episode_limit * 3 candidates)
        candidate_count = episode_limit * 3
        unseen_batch = unseen_pool[unseen_pos : unseen_pos + candidate_count]
        review_batch = review_pool[review_pos : review_pos + candidate_count]

        # Score context for all candidates (batches are already list[VocabularyItem])
        all_candidates = unseen_batch + review_batch
        context_scores: dict[str, float] = await score_context(
            source_text, all_candidates, llm_client
        )

        # Compute final scores for unseen candidates
        unseen_scored: list[tuple[float, Any]] = []
        for item in unseen_batch:
            cs = context_scores.get(item.id, 0.5)
            fs = final_score(item, cs, now)
            unseen_scored.append((fs, item))

        # Compute final scores for review candidates
        review_scored: list[tuple[float, Any]] = []
        for item in review_batch:
            cs = context_scores.get(item.id, 0.5)
            fs = final_score(item, cs, now)
            review_scored.append((fs, item))

        # Allocate based on episode type
        if episode_type == "side":
            pending_item_ids: list[str] = [pw["item_id"] for pw in pending_words]
            target_words, arc_new_ids = allocate_side_episode(
                unseen_scored,
                review_scored,
                pending_item_ids,
                episode_limit=episode_limit,
                arc_new_ids=arc_new_ids,
            )
        else:  # main
            target_words, arc_new_ids = allocate_main_episode(
                unseen_scored,
                review_scored,
                episode_limit=episode_limit,
                arc_new_ids=arc_new_ids,
            )

        # Serialize TargetWord models back to dicts for storage in arc_plan
        episode["target_words"] = [tw.model_dump() for tw in target_words]

        # Advance pool positions by actually consumed counts (not batch size),
        # so unconsumed candidates remain available for subsequent episodes.
        unseen_consumed = sum(1 for tw in target_words if tw.is_new)
        review_consumed = sum(1 for tw in target_words if not tw.is_new)
        unseen_pos += unseen_consumed
        review_pos += review_consumed

    return arc_plan

File: src/services/generation/scheduler.ts
import type { ArcPlan, TargetWord, UserVocabulary, VocabularyItem } from './types';

const EPISODE_LIMIT = 10;

function parseDate(value: string | null | undefined): Date | null {
  if (!value) return null;
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? null : date;
}

function buildPools(
  userVocabulary: UserVocabulary,
  now: Date,
): { unseenPool: VocabularyItem[]; reviewPool: VocabularyItem[] } {
  const unseenPool: VocabularyItem[] = [];
  const reviewPool: VocabularyItem[] = [];

  for (const item of userVocabulary.vocabulary) {
    const lastReview = parseDate(item.fsrs_card.last_review);
    const due = parseDate(item.fsrs_card.due);
    if (!lastReview) {
      unseenPool.push(item);
    } else if (due && due <= now) {
      reviewPool.push(item);
    }
  }

  return { unseenPool, reviewPool };
}

function reorderPool<T extends VocabularyItem>(
  pool: T[],
  pendingOrder: string[],
  pendingIds: Set<string>,
  sortKey: (item: T) => number,
): T[] {
  const poolById = new Map(pool.map((item) => [item.id, item]));
  const pendingItems = pendingOrder
    .map((id) => poolById.get(id))
    .filter((item): item is T => Boolean(item));
  const nonPending = pool
    .filter((item) => !pendingIds.has(item.id))
    .sort((a, b) => sortKey(a) - sortKey(b));
  return [...pendingItems, ...nonPending];
}

function applyPendingOverlay(
  pools: { unseenPool: VocabularyItem[]; reviewPool: VocabularyItem[] },
  pendingWords: Array<{ item_id: string }>,
): { unseenPool: VocabularyItem[]; reviewPool: VocabularyItem[] } {
  const pendingOrder = pendingWords.map((word) => word.item_id);
  const pendingIds = new Set(pendingOrder);
  return {
    unseenPool: reorderPool(
      pools.unseenPool,
      pendingOrder,
      pendingIds,
      (item) => item.chapter_first_seen,
    ),
    reviewPool: reorderPool(
      pools.reviewPool,
      pendingOrder,
      pendingIds,
      (item) => parseDate(item.fsrs_card.due)?.getTime() ?? 0,
    ),
  };
}

function computeUrgency(dueDate: Date, now: Date): number {
  const overdueDays = (now.getTime() - dueDate.getTime()) / 86_400_000;
  if (overdueDays < 0) return 0;
  return Math.min(1, overdueDays / 30);
}

function finalScore(item: VocabularyItem, contextScore: number, now: Date): number {
  if (!item.fsrs_card.last_review) return 0.4 * 0.3 + contextScore * 0.7;
  const due = parseDate(item.fsrs_card.due);
  const urgency = due ? computeUrgency(due, now) : 0;
  return urgency * 0.5 + contextScore * 0.5;
}

function buildTarget(item: VocabularyItem, isNew: boolean): TargetWord {
  return {
    item_id: item.id,
    word: item.word,
    meaning: item.meaning,
    is_new: isNew,
    fsrs_card: item.fsrs_card,
  };
}

function allocateMainEpisode(
  unseenScored: Array<[number, VocabularyItem]>,
  reviewScored: Array<[number, VocabularyItem]>,
  arcNewIds: Set<string>,
): TargetWord[] {
  const targetWords: TargetWord[] = [];
  const unseenSorted = [...unseenScored].sort((a, b) => b[0] - a[0]);
  const reviewSorted = [...reviewScored].sort((a, b) => b[0] - a[0]);

  for (const [, item] of unseenSorted) {
    if (targetWords.filter((tw) => tw.is_new).length >= EPISODE_LIMIT) break;
    if (arcNewIds.has(item.id)) continue;
    targetWords.push(buildTarget(item, true));
    arcNewIds.add(item.id);
  }

  let reviewCount = 0;
  for (const [, item] of reviewSorted) {
    if (reviewCount >= EPISODE_LIMIT) break;
    targetWords.push(buildTarget(item, false));
    reviewCount += 1;
  }

  return targetWords;
}

function allocateSideEpisode(
  unseenScored: Array<[number, VocabularyItem]>,
  reviewScored: Array<[number, VocabularyItem]>,
  pendingItemIds: string[],
  arcNewIds: Set<string>,
): TargetWord[] {
  const pendingSet = new Set(pendingItemIds);
  const pendingUnseen = unseenScored
    .filter(([, item]) => pendingSet.has(item.id))
    .sort((a, b) => b[0] - a[0]);
  const otherUnseen = unseenScored
    .filter(([, item]) => !pendingSet.has(item.id))
    .sort((a, b) => b[0] - a[0]);
  const reviewSorted = [...reviewScored].sort((a, b) => b[0] - a[0]);
  const targetWords: TargetWord[] = [];

  for (const [, item] of [...pendingUnseen, ...otherUnseen]) {
    if (targetWords.filter((tw) => tw.is_new).length >= EPISODE_LIMIT) break;
    if (arcNewIds.has(item.id)) continue;
    targetWords.push(buildTarget(item, true));
    arcNewIds.add(item.id);
  }

  let reviewCount = 0;
  for (const [, item] of reviewSorted) {
    if (reviewCount >= EPISODE_LIMIT) break;
    targetWords.push(buildTarget(item, false));
    reviewCount += 1;
  }

  return targetWords;
}

export async function scheduleVocabulary(
  arcPlan: ArcPlan,
  userVocabulary: UserVocabulary,
  now = new Date(),
): Promise<ArcPlan> {
  const pools = applyPendingOverlay(buildPools(userVocabulary, now), arcPlan.pending_words);
  let unseenPos = 0;
  let reviewPos = 0;
  const arcNewIds = new Set<string>();

  const episodes = arcPlan.episodes.map((episode) => {
    const candidateCount = EPISODE_LIMIT * 3;
    const unseenBatch = pools.unseenPool.slice(unseenPos, unseenPos + candidateCount);
    const reviewBatch = pools.reviewPool.slice(reviewPos, reviewPos + candidateCount);

    const unseenScored = unseenBatch.map(
      (item): [number, VocabularyItem] => [finalScore(item, 0.5, now), item],
    );
    const reviewScored = reviewBatch.map(
      (item): [number, VocabularyItem] => [finalScore(item, 0.5, now), item],
    );

    const targetWords = episode.episode_type === 'side'
      ? allocateSideEpisode(
        unseenScored,
        reviewScored,
        arcPlan.pending_words.map((word) => word.item_id),
        arcNewIds,
      )
      : allocateMainEpisode(unseenScored, reviewScored, arcNewIds);

    unseenPos += targetWords.filter((word) => word.is_new).length;
    reviewPos += targetWords.filter((word) => !word.is_new).length;

    return { ...episode, target_words: targetWords };
  });

  return { ...arcPlan, episodes };
}
